src/record_ts.h

```diff
@@ -0,0 +1,873 @@
+            add_in_param(src_id);
+            add_out_param(dst_id,
+                          this->m_recording.record_variables[src_id].type);
+            uint32_t end = this->m_recording.dependencies.size();
+
+            Operation op;
+            op.type             = OpType::MemcpyAsync;
+            op.dependency_range = std::pair(start, end);
+            op.size             = size;
+            this->m_recording.operations.push_back(op);
+        }
+        {
+            scoped_pause();
+            this->m_internal->memcpy_async(dst, src, size);
+        }
+        if (!paused() && !has_var) {
+            // If we did not know the source variable, this memcpy might be
+            // coming from a `jitc_call_upload` call.
+            // If that is the case, we have to capture the offset buffer.
+            // Since the pointer might be used, for example by an aggregate call
+            // (nested calls), we have to overwrite the RecordVariable.
+            //
+            CallData *call = nullptr;
+            for (CallData *tmp : calls_assembled) {
+                if (tmp->offset == dst) {
+                    call = tmp;
+                    break;
+                }
+            }
+            if (call) {
+                capture_call_offset(dst, size);
+                jitc_log(LogLevel::Debug, "    captured call offset");
+            }
+        }
+    }
+
+    /// Sum over elements within blocks
+    void block_reduce(VarType vt, ReduceOp op, uint32_t size,
+                      uint32_t block_size, const void *in, void *out) override {
+        if (!paused()) {
+            try {
+                record_block_reduce(vt, op, size, block_size, in, out);
+            } catch (...) {
+                record_exception();
+            }
+        }
+        scoped_pause();
+        return this->m_internal->block_reduce(vt, op, size, block_size, in,
+                                              out);
+    }
+
+    void block_prefix_reduce(VarType vt, ReduceOp op, uint32_t size,
+                             uint32_t block_size, bool exclusive, bool reverse,
+                             const void *in, void *out) override {
+        if (!paused()) {
+            try {
+                record_block_prefix_reduce(vt, op, size, block_size, exclusive,
+                                           reverse, in, out);
+            } catch (...) {
+                record_exception();
+            }
+        }
+        scoped_pause();
+        return this->m_internal->block_prefix_reduce(
+            vt, op, size, block_size, exclusive, reverse, in, out);
+    }
+
+    /// Compute a dot product of two equal-sized arrays
+    void reduce_dot(VarType type, const void *ptr_1, const void *ptr_2,
+                    uint32_t size, void *out) override {
+        jitc_log(
+            LogLevel::Warn,
+            "RecordThreadState::reduce_dot(): unsupported function recording!");
```

Comment:
Shouldn't unsupported operations throw an exception? (ideally with as much information as possible to help them find where the unsupported op was triggered).

Reply:
For unsupported operations, you could add a comment explaining why it would be difficult / impossible to support them.